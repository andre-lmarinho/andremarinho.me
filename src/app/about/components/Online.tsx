import { SOCIAL_LINKS } from '@/config/social';

const isMail = (href: string) => href.startsWith('mailto:');

export const Online = () => (
  <section id="online">
    <h2>Online</h2>
    <ul className="flex flex-wrap items-center gap-3">
      {SOCIAL_LINKS.map(({ href, label }) => {
        const external = isMail(href) ? {} : { target: '_blank', rel: 'noopener noreferrer' };
        return (
          <li key={href}>
            <a
              href={href}
              {...external}
              className="flex items-center justify-center rounded-full border border-zinc-200 px-5 py-2 text-sm font-medium text-zinc-700 transition-colors hover:border-zinc-300 hover:bg-zinc-100 focus-visible:ring-2 focus-visible:ring-offset-2 focus-visible:outline-none dark:border-zinc-700 dark:text-zinc-200 dark:hover:border-zinc-500 dark:hover:bg-zinc-900"
            >
              {label}
            </a>
          </li>
        );
      })}
    </ul>
  </section>
);
