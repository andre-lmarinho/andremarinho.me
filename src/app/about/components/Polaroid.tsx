import Image, { type ImageProps } from 'next/image';
import { cn } from '@/utils/cn';

const rotation = {
  '+1': '+rotate-1',
  '-2': '-rotate-2',
  '-3': '-rotate-3',
} as const;

type Props = {
  label: string;
  src: ImageProps['src'];
  rotation: keyof typeof rotation;
};

export const Polaroid = (props: Props) => (
  <div
    className={cn(
      'rounded-lg bg-linear-to-b from-white to-neutral-100 p-4 shadow-2xl select-none',
      'text-center hover:scale-105 hover:rotate-0 dark:from-neutral-200 dark:to-neutral-200',
      'transition-transform ease-out',
      rotation[props.rotation]
    )}
  >
    <div className="relative h-fit w-fit overflow-hidden rounded-sm">
      <Image
        alt={props.label}
        height={240}
        width={240}
        sizes="240px"
        loading="eager"
        placeholder="blur"
        src={props.src}
      />
      <div className="absolute inset-0 shadow-[inset_0_0_4px_rgba(0,0,0,.3)]" />
    </div>
    <p className="text-md mt-3 text-center text-neutral-600 dark:text-neutral-800">{props.label}</p>
  </div>
);
