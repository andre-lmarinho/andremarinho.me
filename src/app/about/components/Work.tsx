import Image from 'next/image';

import { cn } from '@/utils/cn';

type WorkRole = {
  title: string;
  startYear: number;
  endYear?: number | 'Now';
};

type WorkPlace = {
  name: string;
  website: string;
  logo: string;
  roles: WorkRole[];
};

const workPlaces: WorkPlace[] = [
  {
    name: 'Duonorth Studio',
    website: '/studio',
    logo: '/images/work/duonorth.webp',
    roles: [
      {
        title: 'Frontend Developer',
        startYear: 2025,
        endYear: 'Now',
      },
      {
        title: 'Web and WordPress Developer',
        startYear: 2020,
        endYear: 2024,
      },
      {
        title: 'Digital Marketing & Web Consultant',
        startYear: 2017,
        endYear: 2019,
      },
    ],
  },
];

const slugify = (value: string) =>
  value
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/(^-|-$)/g, '');

const toEndValue = (end: WorkRole['endYear']) => {
  if (end === undefined || end === 'Now') return Number.MAX_SAFE_INTEGER;
  return end;
};

const formatPeriod = (role: WorkRole) => {
  const endLabel = role.endYear ?? 'Now';
  return `${role.startYear} - ${endLabel}`;
};

const workEntries = workPlaces
  .flatMap((place) => place.roles.map((role) => ({ place, role })))
  .sort((a, b) => {
    const endDiff = toEndValue(b.role.endYear) - toEndValue(a.role.endYear);
    if (endDiff !== 0) return endDiff;
    return b.role.startYear - a.role.startYear;
  });

export const Work = () => (
  <section id="work" aria-label="Work">
    <h2>Work</h2>
    <ul className="space-y-2">
      {workEntries.map(({ place, role }, index) => {
        const href = place.website;
        const isExternal = /^https?:\/\//.test(href);
        const entryKey = `${slugify(place.name)}-${slugify(role.title)}-${role.startYear}`;
        const showLogo = index === 0;

        return (
          <li key={entryKey} className="h-full">
            <a
              className="group flex gap-4 rounded-xs transition-colors focus-visible:outline focus-visible:outline-offset-2 focus-visible:outline-zinc-500 dark:focus-visible:outline-zinc-400"
              href={href}
              target={isExternal ? '_blank' : undefined}
              rel={isExternal ? 'noreferrer noopener' : undefined}
              aria-label={`${role.title} at ${place.name}`}
            >
              <span
                className={cn(
                  'hidden h-12 w-12 shrink-0 items-center justify-center sm:flex',
                  showLogo
                    ? 'overflow-hidden rounded-xl bg-white/90 shadow-sm dark:bg-zinc-900/70'
                    : 'opacity-0'
                )}
                aria-hidden="true"
              >
                {showLogo ? (
                  <Image
                    src={place.logo}
                    alt=""
                    className="h-full w-full object-contain"
                    width={48}
                    height={48}
                    sizes="48px"
                    priority
                  />
                ) : null}
              </span>
              <div className="flex flex-1 gap-3 sm:gap-4">
                <div className="flex flex-1 flex-col">
                  <span className="leading-tight font-medium text-zinc-900 transition-colors group-hover:text-zinc-600 dark:text-zinc-100 dark:group-hover:text-zinc-400">
                    {role.title}
                  </span>
                  <span className="text-sm text-zinc-600 transition-colors group-hover:text-zinc-500 dark:text-zinc-400 dark:group-hover:text-zinc-500">
                    {place.name}
                  </span>
                </div>
                <span className="ml-auto text-sm font-medium text-zinc-900 transition-colors group-hover:text-zinc-600 dark:text-zinc-100 dark:group-hover:text-zinc-400">
                  {formatPeriod(role)}
                </span>
              </div>
            </a>
          </li>
        );
      })}
    </ul>
  </section>
);
