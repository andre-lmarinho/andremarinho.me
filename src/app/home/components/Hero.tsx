'use client';

import Image from 'next/image';
import React, { useEffect, useState } from 'react';

import cn from '@/utils';

const heroTexts = [
  'I am a Front-End Developer based in Salvador. I create digital experiences that connect design, strategy and business growth.',
  'I run Duonorth Studio, where I turn ideas into products that are fast, meaningful and focused on results.',
] as const;

const waveEmoji = String.fromCodePoint(0x1f44b);

export default function Hero() {
  const [wave, setWave] = useState(false);

  useEffect(() => {
    const id = window.setTimeout(() => setWave(true), 220);
    return () => window.clearTimeout(id);
  }, []);

  return (
    <section id="hero" className="mb-20">
      <div className="flex justify-between gap-8">
        <div>
          <h1 className="flex gap-2">
            <span>Hey! I&apos;m Andre Marinho</span>
            <span
              aria-hidden="true"
              className={cn('inline-block origin-[70%_70%]', wave && 'animate-wave')}
            >
              {waveEmoji}
            </span>
          </h1>
          {heroTexts.map((text) => (
            <p key={text} className="sm:max-w-md">
              {text}
            </p>
          ))}
        </div>
        <Image
          alt="Andre Marinho"
          width={176}
          height={176}
          className="hidden h-44 w-44 transform-gpu rounded-full sm:block"
          src="/images/Me.webp"
          sizes="176px"
          priority
        />
      </div>
    </section>
  );
}
