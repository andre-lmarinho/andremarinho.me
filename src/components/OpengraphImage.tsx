import { readFile } from 'node:fs/promises';
import type { Font } from 'next/dist/compiled/@vercel/og/satori';

type OGProps = {
  description: string;
  title: string;
  url?: string;
};

const assetUrl = (path: string) => new URL(`../../public${path}`, import.meta.url);

const fontSpecs: Array<[Font['name'], NonNullable<Font['weight']>, string]> = [
  ['Inter', 500, '/fonts/opengraph/Inter-Medium.woff'],
  ['Inter', 800, '/fonts/opengraph/Inter-ExtraBold.woff'],
  ['Roboto Mono', 400, '/fonts/opengraph/RobotoMono-Regular.woff'],
];

const fontsPromise: Promise<Font[]> = Promise.all(
  fontSpecs.map(async ([name, weight, path]) => ({
    data: await readFile(assetUrl(path)),
    name,
    style: 'normal',
    weight,
  }))
);

const profileImageSrcPromise = readFile(assetUrl('/images/Me.jpeg')).then(
  (buffer) => `data:image/jpeg;base64,${buffer.toString('base64')}`
);

type OpengraphImageProps = OGProps & { profileImageSrc: string };

function OpengraphImage({ title, description, url, profileImageSrc }: OpengraphImageProps) {
  return (
    <div
      style={{
        backgroundColor: 'white',
        width: '100%',
        height: '100%',
        display: 'flex',
        position: 'relative',
        fontFamily: 'Inter',
        backgroundImage: 'radial-gradient(at 10px 10px, #d4d4d4, #d4d4d4 5%, white 5%)',
        backgroundSize: '20px 20px',
      }}
    >
      <img
        src={profileImageSrc}
        width="115"
        height="115"
        style={{
          borderRadius: '100%',
          position: 'absolute',
          left: 60,
          top: 60,
        }}
      />
      <div
        style={{
          display: 'flex',
          flexDirection: 'column',
          gap: '0.5rem',
          position: 'absolute',
          bottom: 140,
          left: 60,
          width: '65%',
        }}
      >
        <span
          style={{
            fontSize: '5.25rem',
            fontWeight: 700,
            letterSpacing: '-2px',
            lineHeight: '1',
          }}
        >
          {title}
        </span>
        <span
          style={{
            fontSize: '2.5rem',
            fontWeight: 400,
          }}
        >
          {description}
        </span>
      </div>
      <div
        style={{
          position: 'absolute',
          bottom: 60,
          left: 60,
          borderRadius: 100,
          fontFamily: 'RobotoMono',
          fontSize: '1.75rem',
          lineHeight: 1,
          fontWeight: 600,
        }}
      >
        {url ? `andremarinho.me/${url}` : 'andremarinho.me'}
      </div>
    </div>
  );
}

export async function buildOg(props: OGProps) {
  const [fonts, profileImageSrc] = await Promise.all([fontsPromise, profileImageSrcPromise]);
  const el = <OpengraphImage {...props} profileImageSrc={profileImageSrc} />;
  return [el, { fonts }] as const;
}
