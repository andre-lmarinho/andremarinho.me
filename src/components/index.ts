export * from './layout';
export * from './sections';
export * from './visuals';

export { default as ProjectCard } from './ProjectCard';
