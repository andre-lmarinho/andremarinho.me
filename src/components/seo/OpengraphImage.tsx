import { readFile } from 'node:fs/promises';
import type { Font } from 'next/dist/compiled/@vercel/og/satori';
import { siteUrl } from '@/config/metadata';

type Props = {
  description: string;
  title: string;
  url?: string;
};

const buildFontUrl = (path: string) => new URL(`../../../public${path}`, import.meta.url);

type FontConfig = {
  name: Font['name'];
  style: NonNullable<Font['style']>;
  weight: NonNullable<Font['weight']>;
  url: URL;
};

const fontConfigs: FontConfig[] = [
  {
    name: 'Inter',
    style: 'normal',
    weight: 600,
    url: buildFontUrl('/fonts/opengraph/Inter-Medium.woff'),
  },
  {
    name: 'Inter',
    style: 'normal',
    weight: 700,
    url: buildFontUrl('/fonts/opengraph/Inter-ExtraBold.woff'),
  },
  {
    name: 'Roboto Mono',
    style: 'normal',
    weight: 300,
    url: buildFontUrl('/fonts/opengraph/RobotoMono-Regular.woff'),
  },
];

const fontsPromise: Promise<Font[]> = Promise.all(
  fontConfigs.map(async ({ name, style, weight, url }) => ({
    data: await readFile(url),
    name,
    style,
    weight,
  }))
);

export const getFonts = (): Promise<Font[]> => fontsPromise;

export const ogImageSize = {
  width: 1200,
  height: 630,
} as const;

export const ogImageDynamic = 'force-static' as const;

export const createOgImageResponse = async (props: Props) => {
  const { ImageResponse } = await import('next/og');

  return new ImageResponse(<OpengraphImage {...props} />, {
    fonts: await getFonts(),
  });
};

const domain = new URL(siteUrl).host;

const OpengraphImage = ({ description, title, url }: Props) => {
  const displayUrl = url ? `${domain}${url.startsWith('/') ? url : `/${url}`}` : domain;

  return (
    <div
      style={{
        backgroundColor: 'white',
        width: '100%',
        height: '100%',
        display: 'flex',
        position: 'relative',
        fontFamily: 'Inter',
        backgroundImage: 'radial-gradient(at 10px 10px, #d4d4d4, #d4d4d4 5%, white 5%)',
        backgroundSize: '20px 20px',
      }}
    >
      <img
        src="https://andremarinho.me/images/Me.jpeg"
        width="115"
        height="115"
        style={{
          borderRadius: '100%',
          position: 'absolute',
          left: 60,
          top: 60,
        }}
      />
      <div
        style={{
          display: 'flex',
          flexDirection: 'column',
          gap: '0.5rem',
          position: 'absolute',
          bottom: 140,
          left: 60,
          width: '65%',
        }}
      >
        <span
          style={{
            fontSize: '5.25rem',
            fontWeight: 700,
            letterSpacing: '-2px',
            lineHeight: '1',
          }}
        >
          {title}
        </span>
        <span
          style={{
            fontSize: '2.5rem',
            fontWeight: 400,
          }}
        >
          {description}
        </span>
      </div>
      <div
        style={{
          position: 'absolute',
          bottom: 60,
          left: 60,
          borderRadius: 100,
          fontFamily: 'RobotoMono',
          fontSize: '1.75rem',
          lineHeight: 1,
          fontWeight: 600,
        }}
      >
        {displayUrl}
      </div>
    </div>
  );
};

export default OpengraphImage;
