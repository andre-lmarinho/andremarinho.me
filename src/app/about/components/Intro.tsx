import localFont from 'next/font/local';

import { SmartLink as A } from '@/components/Link';
import { PageTitle } from '@/components/Heading';
import { cn } from '@/utils/cn';
import { Polaroid } from './Polaroid';
import dog from './images/doggy.jpg';
import beach from './images/beach.jpg';
import whitesand from './images/whitesand.jpg';

const gloria = localFont({
  src: [
    {
      path: '../../../../public/fonts/gloria-hallelujah.woff2',
      style: 'normal',
    },
  ],
  weight: '400',
  display: 'swap',
  declarations: [
    {
      prop: 'unicode-range',
      value:
        'U+00??, U+0131, U+0152-0153, U+02bb-02bc, U+02c6, U+02da, U+02dc, U+2000-206f, U+2074, U+20ac, U+2122, U+2191, U+2193, U+2212, U+2215, U+feff, U+fffd',
    },
  ],
});

export const Intro = () => (
  <section id="about" className="mt-6">
    <PageTitle>About me</PageTitle>
    <div className="mb-6 space-y-5">
      <p>
        <em>Hey!</em> üëãüèº
      </p>
      <p>
        I&apos;m Andr√©, a Front End Developer based in Salvador, that loves to code and build
        products with a delightful user experience.
      </p>

      <p>
        Currently working at <A href="/studio">Duonorth Studio</A> as a Front End Developer.
      </p>

      <p>
        I love working in between product, engineering and developer experience. Some things that
        make me excited are TypeScript, Open Source, CI & CD systems, simplicity, automating things
        and building meaningful user-centric products.
      </p>

      <p>
        I was born and raised in salvador next to the beach and I&apos;m studying Computer Science
        at <A href="https://github.com/ossu/computer-science">OSSU</A> with some Harvard&apos;s
        courses like{' '}
        <A href="https://certificates.cs50.io/d7d5d5d9-8e76-4b4d-b324-9ed0fee6f9f5.pdf">CS50</A> and{' '}
        <A href="https://github.com/andre-lmarinho/cs50w">CS50W</A>.
      </p>

      <p>
        Outside work, I&apos;m into training, building small tools, and{' '}
        <A href="https://travel-planner-orpin.vercel.app/">exploring new places</A> with my family.
      </p>
    </div>

    <div className={cn('flex gap-0 py-6 lg:flex-wrap lg:gap-6', gloria.className)}>
      <Polaroid label="Doggy ‚ù£Ô∏è" src={dog} rotation="-2" />
      <Polaroid label="Beach '22 üèñÔ∏è" src={beach} rotation="+1" />
      <div className="hidden sm:block">
        <Polaroid label="White peace üßòüèº" src={whitesand} rotation="-3" />
      </div>
    </div>
  </section>
);
