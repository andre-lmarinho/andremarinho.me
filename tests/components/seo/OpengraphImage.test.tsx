import { render, screen } from '@testing-library/react';

import { OpengraphImage } from '@/components/OpengraphImage';

describe('OpengraphImage', () => {
  it('matches the expected structure for the default card', () => {
    const { container } = render(
      <OpengraphImage title="AndrÃ© Marinho" description="This is a description test." />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the scoped url when provided', () => {
    render(<OpengraphImage title="About me" description="Learn about Andre" url="about" />);

    expect(screen.getByText('andremarinho.me/about')).toBeInTheDocument();
  });
});
