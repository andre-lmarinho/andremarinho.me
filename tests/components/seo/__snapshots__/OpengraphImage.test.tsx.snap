// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`OpengraphImage matches the expected structure for the default card 1`] = `
<div
  style="background-color: white; width: 100%; height: 100%; display: flex; position: relative; font-family: Inter; background-image: radial-gradient(at 10px 10px, #d4d4d4, #d4d4d4 5%, white 5%); background-size: 20px 20px;"
>
  <img
    height="115"
    src="https://andremarinho.me/images/Me.jpeg"
    style="border-radius: 100%; position: absolute; left: 60px; top: 60px;"
    width="115"
  />
  <div
    style="display: flex; flex-direction: column; gap: 0.5rem; position: absolute; bottom: 140px; left: 60px; width: 65%;"
  >
    <span
      style="font-size: 5.25rem; font-weight: 700; letter-spacing: -2px; line-height: 1;"
    >
      Andr√© Marinho
    </span>
    <span
      style="font-size: 2.5rem; font-weight: 400;"
    >
      This is a description test.
    </span>
  </div>
  <div
    style="position: absolute; bottom: 60px; left: 60px; border-radius: 100px; font-family: RobotoMono; font-size: 1.75rem; line-height: 1; font-weight: 600;"
  >
    andremarinho.me
  </div>
</div>
`;
